rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,578 @@
+    if (format == "u8") THROW("uint64_t dtype not supported in rocAL");
+    if (format == "i1") return RocalTensorDataType::INT8;
+    if (format == "i2") return RocalTensorDataType::INT16;
+    if (format == "i4") return RocalTensorDataType::INT32;
+    if (format == "i8") THROW("int64_t dtype not supported in rocAL");
+    if (format == "f2")
+#if defined(AMD_FP16_SUPPORT)
+        return RocalTensorDataType::FP16;
+#else
+        THROW("FLOAT16 type tensor not supported")
+#endif
+    if (format == "f4") return RocalTensorDataType::FP32;
+    if (format == "f8") THROW("double dtype not supported in rocAL");
+    THROW("Unknown Numpy dtype string");
+}
+
+inline void NumpyDataReader::skip_spaces(const char*& ptr) {
+    while (::isspace(*ptr))
+        ptr++;
+}
+
+template <size_t N>
+void NumpyDataReader::skip_char(const char*& ptr, const char (&what)[N]) {
+    if (strncmp(ptr, what, N - 1))
+        THROW("Found wrong symbol during parsing");
+    ptr += N - 1;
+}
+
+template <size_t N>
+bool NumpyDataReader::try_skip_char(const char*& ptr, const char (&what)[N]) {
+    if (!strncmp(ptr, what, N - 1)) {
+        ptr += N - 1;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+template <size_t N>
+void NumpyDataReader::skip_field(const char*& ptr, const char (&name)[N]) {
+    skip_spaces(ptr);
+    skip_char(ptr, "'");
+    skip_char(ptr, name);
+    skip_char(ptr, "'");
+    skip_spaces(ptr);
+    skip_char(ptr, ":");
+    skip_spaces(ptr);
+}
+
+template <typename T = int64_t>
+T NumpyDataReader::parse_int(const char*& ptr) {
+    char* out_ptr = const_cast<char*>(ptr);  // strtol takes a non-const pointer
+    T value = static_cast<T>(strtol(ptr, &out_ptr, 10));
+    if (out_ptr == ptr)
+        THROW("Parse error: expected a number.");
+    ptr = out_ptr;
+    return value;
+}
+
+std::string NumpyDataReader::parse_string(const char*& input, char delim_start, char delim_end) {
+    if (*input++ != delim_start)
+        THROW("Expected \'" + std::to_string(delim_start) + "\'");
+    std::string out;
+    for (; *input != '\0'; input++) {
+        if (*input == '\\') {
+            switch (*++input) {
+                case '\\':
+                    out += '\\';
+                    break;
+                case '\'':
+                    out += '\'';
+                    break;
+                case '\t':
+                    out += '\t';
+                    break;
+                case '\n':
+                    out += '\n';
+                    break;
+                case '\"':
+                    out += '\"';
+                    break;
+                default:
+                    out += '\\';
+                    out += *input;
+                    break;
+            }
+        } else if (*input == delim_end) {
+            break;
+        } else {
+            out += *input;
+        }
+    }
+    if (*input++ != delim_end)
+        THROW("Expected \'" + std::to_string(delim_end) + "\'");
```

Comment:
Mention in which reader you are throwing error or warning. Please check for other instances too.